docs/framework-design.md

```diff
@@ -0,0 +1,108 @@
+
+### Usage of the repository
+
+The [`mbedtls-framework`](https://github.com/Mbed-TLS/mbedtls-framework) repository is consumed by:
+
+* Mbed TLS 3.6 LTS (`mbedtls-3.6` branch);
+* Mbed TLS 4.x (to become the `development` branch, once 4.x development has) sufficiently advanced;
+* TF-PSA-Crypto;
+* as well as future Mbed TLS and TF-PSA-Crypto LTS branches.
+
+In each consuming repository, the `mbedtls-framework` repository appears as a Git submodule located at the path `/framework`.
+
+### Requirements for the framework repository
+
+#### Framework repository versioning
+
+The framework repository is not versioned: projects are only supposed to consume the tip of the `main` branch. There are no tagged releases. However, each release of a consuming repository will designate a specific commit of the framework repository (this behavior is built into Git submodules), which can be tagged accordingly.
+
+At any point in time, each consuming repository requires a specific commit in the framework repository. Moving a consuming repository to the tip of the framework repository is a manual action. As a consequence, breaking changes are possible: they will not break any actual commit, they would only prevent the consuming repostiory to update its submodule version to the tip of the framework repository.
+
+However, breaking changes are still problematic. Breaking changes in the framework repository require the affected consuming repositories to fully adapt to the changes when they want to gain access to any new features in the framework repository. Breaking changes in a consuming repository that affect a feature that is consumed by the consuming repository (e.g. internal library functions called by test helper functions) have the same effect.
+
+To facilitate parallel development, major changes should avoid breaking existing code and should provide a transition period. For example, if a function needs a new argument, define a new function with a new name, start using the new fuction, and later remove the old function.
+
+### Requirements for consuming repositories
+
+Consuming repositories must have the framework repository as a Git submodule for development work and CI scripts. For merely using the content of the consuming repository, in particular building and running tests, it is enough to have the content of the framework repository, for example from a zip file.
+
+Release archives must include the content of the framework repository.
+
+Without the content of the framework repository, there is no expectation that the consuming repository is usable. In particular, we do not expect the following to work:
+
+* Generating configuration-independent files (e.g. `make generated_files`), not even the ones in the `library` directory.
+* `make lib` (with GNU make or CMake) from a pristine checkout.
+* `make test` (even if all the tests have been built).
```

Comment:
I think this could be simplified and reworded a bit to make the main points clear:
* Consuming repositories MUST have the framework repo as a submodule
* They MAY rely on the framework git submodule during development
* They MAY rely on the contents of the framework submodule for any build
   * [Subject to previous comment] EXCEPT when configuration-independent generated files are already present, in which case the it MUST be possible to build the library only without the `framework` submodule.
   * So long as the contents match those of the framework commit pointed to by the corresponding submodule in git. Consuming repositories are not required to consume the contents of an arbitrary `framework` version.

Reply:
I don't really understand your presentation of information, but I think I understand your concern with the current wording, so I've completely reworded this section in a different way.